sophosendpointprotection_connector.py

```diff
@@ -0,0 +1,829 @@
+            ret_val, resp_json = self._make_rest_call(url, action_result, headers, params, data, json, method)
+
+        self.save_progress(("Response in JSON: {}".format(str(resp_json))))
+
+        if phantom.is_fail(ret_val):
+            return (action_result.get_status(), None)
+        else:
+            return (
+             phantom.APP_SUCCESS, resp_json)
+
+    def _get_token(self, action_result, from_action=False):
+        data = {
+            'client_id': self._client_id,
+            'client_secret': self._client_secret,
+            'scope': 'token',
+            'grant_type': 'client_credentials'
+        }
+        # self.save_progress("Data: {}".format(json.dumps(data)))
+        headers = {'Content-Type': 'application/x-www-form-urlencoded', 'Accept': 'application/json'}
+        url = ("{}".format(JWT_TOKEN_ENDPOINT))
+        self.save_progress(("Fetching the JWT token"))
+        self.save_progress(("Hitting the URL for JWT token: {}".format(url)))
+        ret_val, resp_json = self._make_rest_call(url, action_result, headers=headers, data=data, method='post')
+        self.save_progress(("Response in JSON: {}".format(resp_json)))
+        self.save_progress(("Return value: {}".format(ret_val)))
+        if not ret_val:
+            return self.set_status(phantom.APP_ERROR, "Token not found")
+
+        self.save_progress(("Saving to state"))
+        self._state[SOPHOS_JWT_JSON] = resp_json
+        self._JWT_token = resp_json[SOPHOS_JWT_TOKEN]
+        self.save_progress("Got the token: {}".format(self._JWT_token))
+
+        # Getting the X-Tenant-ID
+        data = {}
+        headers = {'Authorization': 'Bearer {}'.format(self._JWT_token), 'Accept': 'application/json'}
+        url = ("{}".format(WHOAMI_ENDPOINT))
+        self.save_progress(("Fetching the partner token"))
+        ret_val, resp_json = self._make_rest_call(url, action_result, headers=headers, data=data, method='get')
+
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        self._state[SOPHOS_PT_JSON] = resp_json
+        self._partner_token = resp_json[SOPHOS_PT_TOKEN]
+        self._id_type = resp_json["idType"]
+        # idType = partner | organization | tenant
+        if resp_json["idType"] == 'tenant':
+            self._base_url = resp_json[SOPHOS_PT_API_HOSTS][SOPHOS_PT_DATA_REGION_URL]
+        else:
+            self._base_url = resp_json[SOPHOS_PT_API_HOSTS][SOPHOS_PT_GLOBAL_URL]
+        self.save_state(self._state)
+        self.save_progress("Got the partner token")
+        return phantom.APP_SUCCESS
+
+    """
+    Get Partner Token.
+    """
+    # def _get_partner_token(self,jwt_token):
+    #     self.save_progress("Getting the partner token")
+    #     headers = {'Authorization': 'Bearer '.format(jwt_token)}
+    #     payload = {}
+    #     response = requests.request("POST", WHOAMI_ENDPOINT, headers=headers, data = payload)
+    #     return json.loads(response.text)['id']
+
+    """
```

Comment:
Docstring should be moved below method definition.

Reply:
Moved docstring below method definition.